internal/internal_worker_base.go

```diff
@@ -332,61 +347,100 @@ func (bw *baseWorker) runPoller() {
 	defer bw.stopWG.Done()
 	bw.metricsHandler.Counter(metrics.PollerStartCounter).Inc(1)
 
+	ctx, cancelfn := context.WithCancel(context.Background())
+	reserveChan := make(chan *SlotPermit)
+
 	for {
+		bw.stopWG.Add(1)
+		go func() {
+			defer bw.stopWG.Done()
+			s, err := bw.slotSupplier.ReserveSlot(ctx, &bw.options.slotReservationData)
+			if err != nil {
+				if !errors.Is(err, context.Canceled) {
+					bw.logger.Error(fmt.Sprintf("Error while trying to reserve slot: %v", err))
+				} else {
+					close(reserveChan)
+				}
+				return
+			}
+			select {
+			case reserveChan <- s:
+			case <-ctx.Done():
+				bw.releaseSlot(s, "worker shutting down")
+			}
+		}()
+
 		select {
 		case <-bw.stopCh:
+			cancelfn()
```

Comment:
It is a tad safer and clearer to put this as `defer cancelfn()` right below where it is created (obviously doesn't matter that much in here, but is safer for future code edits that may come through)